### ISSUE TITLE
English instead of German number seperator signs for German culture

### ISSUE THREAD

**Issue by MaximilianKodat** [EPPlusSoftware/EPPlus#1483]
### EPPlus usage

Commercial use (I have a commercial license)

### Environment

Windows

### Epplus version

7.1.3

### Spreadsheet application

Excel

### Description

Hello,
we (AMANA) are facing some problems with number seperator signs for German culture. Thousands seperator in German is "." instead of the English "," and the decimal seperator in German is "," instead of the English ".", so basically it's swapped.
We have tried to fix this and also created a unit test for this behavior. I will add the test to the end of this description.

Please note that this UT has some commented out asserts which had additional problems with the decimal seperator.

```
        [TestMethod]
        public void IssueMissingDecimalsTextFormular()
        {
            //Issue: TEXT-formular deletes decimals in german format
#if Core
            var dir = AppContext.BaseDirectory;
            dir = Directory.GetParent(dir).Parent.Parent.Parent.FullName;
#else
            var dir = AppDomain.CurrentDomain.BaseDirectory;
#endif
            using var exlPackage = new ExcelPackage(new FileInfo(Path.Combine(dir, "Workbooks", "Textformat.xlsx")));

            var previousCi = ExcelWorkbook.Culture;
            ExcelWorkbook.Culture = new CultureInfo("de-DE");
            exlPackage.Workbook.Calculate();
            ExcelWorkbook.Culture = previousCi;

            Assert.AreEqual("292.336,30 €", exlPackage.Workbook.Worksheets[0].Cells["A1"].Text);
            Assert.AreEqual("292336,300000 €", exlPackage.Workbook.Worksheets[0].Cells["A2"].Value);
            Assert.AreEqual("292.336 €", exlPackage.Workbook.Worksheets[0].Cells["A3"].Value);

            Assert.AreEqual("292.336,30 €", exlPackage.Workbook.Worksheets[0].Cells["A5"].Value);
            Assert.AreEqual("-292336-- €", exlPackage.Workbook.Worksheets[0].Cells["A6"].Value);
            //Assert.AreEqual("-292336,--€", exlPackage.Workbook.Worksheets[0].Cells["A6"].Value);
            Assert.AreEqual("233.127,25 €)", exlPackage.Workbook.Worksheets[0].Cells["A7"].Value);
            Assert.AreEqual("-233127--€)", exlPackage.Workbook.Worksheets[0].Cells["A8"].Value);
            //Assert.AreEqual("-233127,--€)", exlPackage.Workbook.Worksheets[0].Cells["A8"].Value);
            Assert.AreEqual("0,00 €", exlPackage.Workbook.Worksheets[0].Cells["A9"].Value);
            Assert.AreEqual("--- €", exlPackage.Workbook.Worksheets[0].Cells["A10"].Value);
            //Assert.AreEqual("-,-- €", exlPackage.Workbook.Worksheets[0].Cells["A10"].Value);
            Assert.AreEqual("0,00 €)", exlPackage.Workbook.Worksheets[0].Cells["A11"].Value);
            Assert.AreEqual("---€)", exlPackage.Workbook.Worksheets[0].Cells["A12"].Value);
            //Assert.AreEqual("-,--€)", exlPackage.Workbook.Worksheets[0].Cells["A12"].Value);
            Assert.AreEqual("1.027,60 €", exlPackage.Workbook.Worksheets[0].Cells["A13"].Value);
            Assert.AreEqual("-1028-- €)", exlPackage.Workbook.Worksheets[0].Cells["A14"].Value);
            //Assert.AreEqual("-1028,--€)", exlPackage.Workbook.Worksheets[0].Cells["A14"].Value);
            Assert.AreEqual("445,58 €)", exlPackage.Workbook.Worksheets[0].Cells["A15"].Value);
            Assert.AreEqual("-446-- €)", exlPackage.Workbook.Worksheets[0].Cells["A16"].Value);
            //Assert.AreEqual("-446,--€)", exlPackage.Workbook.Worksheets[0].Cells["A16"].Value);
            Assert.AreEqual("0,00 €", exlPackage.Workbook.Worksheets[0].Cells["A17"].Value);
            Assert.AreEqual("0,00 €)", exlPackage.Workbook.Worksheets[0].Cells["A18"].Value);
            Assert.AreEqual("--- €)", exlPackage.Workbook.Worksheets[0].Cells["A19"].Value);
            //Assert.AreEqual("-,--€)", exlPackage.Workbook.Worksheets[0].Cells["A19"].Value);
            Assert.AreEqual("--- €", exlPackage.Workbook.Worksheets[0].Cells["A20"].Value);
            //Assert.AreEqual("-,--€", exlPackage.Workbook.Worksheets[0].Cells["A20"].Value);
}
```
[Textformat.xlsx](https://github.com/user-attachments/files/15805729/Textformat.xlsx)

**Comment by MaximilianKodat** [comment]
[Textformat.xlsx](https://github.com/user-attachments/files/15804259/Textformat.xlsx)

**Comment by JulianGrayMunich** [comment]
Hi.. 
I live and work in both Germany and the UK. My software gets deployed in both countries. My main qualifications are in Engineering, secondary in software development, so I give these comments  to help for what it is worth.

I experienced the same problems as you, but also wrt time analysis (I work with data/time series). This caused me endless grief. The initial issue I faced was that all my unit tests worked in Germany, but when I deployed in the UK my software failed. Sometimes it appeared to work, but it computed incorrectly which was a nightmare. The 2 main causes: US date vs UK date format (switching months and days around), comma seperator vs decimal point seperator.

I now always install "using System.Globalization;" and use "CultureInfo.InvariantCulture" wherever necessary - to be clear, I code in C#.

I run my dates through a verification method both when extracting off a worksheet, or when reading date/time from the server. Similarly with numeric data which I may be extracting from, or writing to the worksheet. I adopt the English/UK culture of decimal point seperator and ddMMYYYY  or YYYYMMdd when working with EPPLus.

If you deploy in South Africa you will face a mixture of both cultures (decimal point and decimal comma as seperators) to add to your misery. 

Strictly speaking my comments do not apply to your unit test, so maybe you must read them and then delete, but I hope this makes you aware of other issues that you may face, issues which nearly cost me a consulting contract. 

Julian

**Comment by JanKallman** [comment]
The supplied test code seems to reference a property that does not exist in EPPlus ...  `ExcelWorkbook.Culture`

**Comment by MaximilianKodat** [comment]
Hi Julian,
thank you for your response. We already tried to work around this issue in the past and your suggestions are helpful. This issue was created to see, if there is a way to do this directly with EPP or if there is a plan to fix/avoid this culture problem.

Hi Jan,
the mentioned code is just a small remaining part of a legacy fix we created for this issue. When removed, the test should also fail as expected. We tried to set the cultures manually and added some code to replace the seperators when encountering German culture. My question would be, if there is a way to get the correct results for sheet with German culture?

**Comment by JanKallman** [comment]
Looks like the `Text` function replaces the decimal separator, but not the thousand separator, causing this issue. I will try to provide a fix.

**Comment by oharabati** [comment]
Hi Jan,
thank you for the prompt response and for looking into the issue :)

We (AMANA) can create a pull request with our solution for this problem, which we used in the latest few years in our forked version of EPPLus. Would it be helpfull for you?

**Comment by JanKallman** [comment]
I have added a generic fix for this that will be released in EPPlus 7.2.

**Comment by MaximilianKodat** [comment]
Thank you :)
